### PR TITLE
Add support for Externalizable in configuration cache

### PR DESCRIPTION
**Pull request by abstratt**
Fixes: #28461

### CODE REVIEW

**Comment by abstratt** [comment]
@bot-gradle test RFN

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/Contexts.kt [hunk 1]
```diff
@@ -399,9 +399,8 @@ abstract class AbstractIsolateContext<T>(
     }
 
     override fun push(owner: IsolateOwner, codec: Codec<Any?>) {
-        contexts.add(0, Pair(currentIsolate, currentCodec))
+        push(codec)
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/Contexts.kt:402]: 👍

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/BindingsBackedCodec.kt [hunk 2]
```diff
@@ -33,6 +33,12 @@ import org.gradle.internal.serialize.Serializer
 import kotlin.reflect.KClass
 
 
+/**
+ * An implementation of the Codec protocol that (based on a [Binding.tag]) chooses and delegates
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/BindingsBackedCodec.kt]: The binding (a tagged codec) is chosen based on the availability of a `Binding.encoding` for the value being encoded. This is basically implemented as a predicate dispatching on the value type, first available `Binding.encoding` wins and its `Binding.tag` is recorded in the output stream so decoding can be implemented via a fast array lookup.

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/BindingsBackedCodec.kt [hunk 3]
```diff
@@ -91,8 +97,14 @@ data class Binding(
 }
 
 
+/**
+ * An object that can determine, for a given type, whether it can encode instances of that type, and which specific encoding to use.
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/BindingsBackedCodec.kt]: Nailed it! 😄

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/jos/JavaObjectSerializationCodec.kt [hunk 4]
```diff
@@ -82,6 +83,12 @@ class JavaObjectSerializationCodec(
     override suspend fun ReadContext.decode(): Any =
         decodePreservingIdentity { id ->
             when (readEnum<Format>()) {
+                Format.Externalizable -> {
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/jos/JavaObjectSerializationCodec.kt:86]: ❓ Did you consider making it a top-level encoder instead?

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/package-info.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * This subsystem implements serialization of configuration state.
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/package-info.java]: 👏

File: platforms/core-configuration/configuration-cache/src/test/kotlin/org/gradle/configurationcache/serialization/codecs/JavaObjectSerializationCodecTest.kt [hunk 6]
```diff
@@ -133,30 +131,44 @@ class JavaObjectSerializationCodecTest : AbstractUserTypeCodecTest() {
...
 
     @Test
-    fun `Externalizable problems link to the Java serialization section`() {
-        val problems = serializationProblemsOf(ExternalizableBean())
-        assertThat(problems.size, equalTo(1))
-        assertThat(
-            problems[0].documentationSection,
-            equalTo(NotYetImplementedJavaSerialization)
-        )
+    fun `can handle Externalizable beans`() {
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/test/kotlin/org/gradle/configurationcache/serialization/codecs/JavaObjectSerializationCodecTest.kt:134]: 🧪 It'd be good to prove `Externalizable` identities are preserved as well

**Review by bamboo — CHANGES_REQUESTED** [review]

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/jos/JavaObjectSerializationCodec.kt (hunk 4 above)
Comment (abstratt) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/jos/JavaObjectSerializationCodec.kt:86]: I did consider that. I figured `Externalizable` is closely related to the rest of the Java Object Serialization (and also a `Serializable`, so there would be benefit in keeping it next to the support for other kinds of `Serializable` types.

File: platforms/core-configuration/configuration-cache/src/test/kotlin/org/gradle/configurationcache/serialization/codecs/JavaObjectSerializationCodecTest.kt (hunk 6 above)
Comment (abstratt) [reply platforms/core-configuration/configuration-cache/src/test/kotlin/org/gradle/configurationcache/serialization/codecs/JavaObjectSerializationCodecTest.kt:134]: I had a question about that: should we ensure it is preserved, or ensure it is *not* preserved? I was assuming/suspecting externalizable objects were not reused, but I was wrong about that.

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/package-info.java (hunk 5 above)
Comment (abstratt) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/package-info.java]: This change was coming from a cherry picked commit that ended up being merged to master. 

Please see https://github.com/gradle/gradle/pull/28563 for PR that contributed the contents for this file.

File: platforms/core-configuration/configuration-cache/src/test/kotlin/org/gradle/configurationcache/serialization/codecs/JavaObjectSerializationCodecTest.kt (hunk 6 above)
Comment (abstratt) [reply platforms/core-configuration/configuration-cache/src/test/kotlin/org/gradle/configurationcache/serialization/codecs/JavaObjectSerializationCodecTest.kt:134]: Done!

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/jos/JavaObjectSerializationCodec.kt (hunk 4 above)
Comment (bamboo) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/jos/JavaObjectSerializationCodec.kt:86]: I see. I'm ambivalent about the runtime cost of the setup, namely, the double `instanceof` check on every `Serializable` instance plus, and more importantly, the BuildTree scoped cache entry in `JavaSerializationEncodingLookup`. I'm leaning slightly towards a top-level codec.

Comment (bamboo) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/jos/JavaObjectSerializationCodec.kt:86]: There's also a data cost to the unnecessary indirection.

**Review by bamboo — APPROVED** [review]
💯 !

**Comment by abstratt** [comment]
@bot-gradle test RFN

**Comment by abstratt** [comment]
@bot-gradle test and merge

**Pull request merged**